chore(health): rewrite health probe doc (#750)
kumahq/kuma-website: docs/docs/dev/policies/service-health-probes.md

# Service Health Probes

Kuma is able to track the status of the Envoy proxy and the underlying app.

> Compared to `HealthCheck` policies Health Probes have the following advantages:

Reviewer: ```suggestion
Compared to `HealthCheck` policies, Health Probes have the following advantages:
```

- knowledge about health is propagated back to `kuma-cp` and is visible in both in Kuma GUI and Kuma CLI
- scalable with thousands of Dataplanes

> Unlike `HealthCheck` policy Health Probes:

Reviewer: ```suggestion
Unlike `HealthCheck` policies, Health Probes:
```

without a comma it's not clear which are the two noun phrases

        kuma.io/protocol: http
```

This `health.ready` status is intended to show the status of the endpoint itself.
It is set differently depending on the environment ([Kubernetes](#kubernetes) or [Universal](#universal-probes)).
It's treated the same way regardless of the environment:

- if `health.ready` is true or `health` section is missing - Kuma considers the inbound as healthy and includes it 
  into load balancing set.
- if `health.ready` is false -  Kuma doesn't include the inbound into the load balancing set.

Also, `health.ready` is used to compute the status of the Dataplanes and Service. You can see these statuses both in Kuma GUI and Kuma CLI:

> Dataplane health follows these rules:

Reviewer: IMO either `Dataplane` or just data plane proxy (there are more instances in the rest of the PR)

> data plane proxy health follows these rules:

Reviewer: ```suggestion
Data plane proxy health follows these rules:
```
sorry :see_no_evil:

- if all dataplanes are `Online` then the service is `Online`
- if no dataplane is `Online` then the service is `Offline`
- if at least one of the dataplanes is `Offline` are `Partially degraded` then the service is `Partially degraded`

## Kubernetes

Kuma leverages container statuses within the pod to evaluate `dataplane.inbound.health` and will follow these rules:

- If the sidecar container is not ready then all inbound will have `health=false`.
- If the side container is ready then the health of the inbound will be whatever is the readiness of the container which exposes the port used by the inbound. 

### Virtual Probes

For better lifecycle management Kubernetes has [readiness, liveness and startup probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/).

When using mTLS the container ports are not available outside the pod and Kubernetes will fail to check the probes.
To work around this Kuma generates a special non-mTLs listener and overrides the probe definitions in the Pod to proxy them through this sidecar listener.

> This is called *Virtual probes*.

Reviewer: ```suggestion
These are called *Virtual probes*.
```

::: warning

> This feature currently only support `httpGet` probes.

Reviewer: ```suggestion
This feature currently only supports `httpGet` probes.
```

  periodSeconds: 3
```

Where `9000` is a default virtual probe port, this is configurable:

:::: tabs :options="{ useUrlFragment: false }"
::: tab "control plane config"
With the config yaml:
```yaml
runtime:
  kubernetes:
    injector:
      virtualProbesPort: 19001
```

or environment variable: `KUMA_RUNTIME_KUBERNETES_VIRTUAL_PROBES_PORT=19001`
:::
::: tab "Pod annotation"
With the Pod annotation:
```yaml
annotations:
  kuma.io/virtual-probes-port: 19001
```
:::
::::

> You can also disable Kuma's probe virtualization

Reviewer: ```suggestion
You can also disable Kuma's probe virtualization:
```

:::: tabs :options="{ useUrlFragment: false }"
::: tab "control plane config"

With the config yaml:
```yaml
runtime:
  kubernetes:
    injector:
      virtualProbesEnabled: false
```
or environment variable: `KUMA_RUNTIME_KUBERNETES_VIRTUAL_PROBES_ENABLED=false`
::: tab "Pod annotation"
With the Pod annotation:
```yaml
annotations:
  kuma.io/virtual-probes: disabled
```
:::
::::

## Universal probes

On Universal there is no single standard for probing the service.

> For health checking the dataplane status on Universal Kuma is using Envoy's Health Discovery Service (HDS).

Reviewer: ```suggestion
For health checking, the dataplane status on Universal Kuma uses Envoy's Health Discovery Service (HDS).
```

        timeout: 2s # optional (default value is taken from KUMA_DP_SERVER_HDS_CHECK_TIMEOUT)
        interval: 1s # optional (default value is taken from KUMA_DP_SERVER_HDS_CHECK_INTERVAL)
        healthyThreshold: 1 # optional (default value is taken from KUMA_DP_SERVER_HDS_CHECK_HEALTHY_THRESHOLD)
        unhealthyThreshold: 1 # optional (default value is taken from KUMA_DP_SERVER_HDS_CHECK_UNHEALTHY_THRESHOLD)
        tcp: {}
      tags:
        kuma.io/service: backend
        kuma.io/protocol: http
```

If there is a `serviceProbe` configured for the inbound, Kuma will automatically fill the `inbound.health` section and update it 

> with interval equal to `KUMA_DP_SERVER_HDS_REFRESH_INTERVAL`.

Reviewer: ```suggestion
with the interval equal to `KUMA_DP_SERVER_HDS_REFRESH_INTERVAL`.
```

Alternatively, it's possible to omit a `serviceProbe` section and develop custom automation that periodically updates the `health` of the inbound.

> If grpc stream with Envoy is disconnected then Kuma considers this proxy as offline.

Reviewer: ```suggestion
If the grpc stream with Envoy is disconnected then Kuma considers this proxy offline.
```

> It will however, still advertise inbounds as their latest health was.

Reviewer: ```suggestion
It will, however, still advertise inbounds using the final update on their health before disconnection.
```

If I understood correctly.

> This is to avoid connection issues between `kuma-cp` and `kuma-dp` to block cross DP traffic.

Reviewer: ```suggestion
This is to avoid connection issues between `kuma-cp` and `kuma-dp` blocking data plane traffic.
```
